### PR TITLE
[refactor] #3887: Hide an internal structure of the versioned structs.

### LINKED ISSUE #3887: Prune usage of V1 structs from codebase
Let's see to what extent can we rewrite the code (source code or tests) that destructures versioned enum into a concrete type:
- for example `VersionedSignedBlock::V1` must not be used except in `data_model`
- no code outside of `data_model` should depend on a specific version

### PR DESCRIPTION
**Pull request by Stukalov-A-M**
## Description

Encapsulated the next structs:
* SignedTransactionV1
* SignedBlockV1

The next structs were changed:
* ValidBlock(pub(super) Signedblock)
* CommitedBlock(pub(super) ValidBlock)

### Linked issue

Closes #3887 

### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [x] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [x] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: client/src/client.rs
Comment (mversic) [inline client/src/client.rs:1664]:
```suggestion
            tx.set_creation_time(tx1.creation_time().as_millis().try_into().unwrap());
```

File: core/src/tx.rs
Comment (mversic) [inline core/src/tx.rs:146]:
```suggestion
```

remove this

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:265]:
```suggestion
    pub fn creation_time(&self) -> Duration {
```

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:272]: remove this function

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:395]:
```suggestion
    pub fn creation_time(&self) -> Duration {
```

File: core/src/block.rs
Comment (mversic) [inline core/src/block.rs:648]: I suggest you also remove this and make proxy functions

File: core/src/block.rs
Comment (mversic) [inline core/src/block.rs:249]: likewise

File: data_model/src/block.rs
Comment (mversic) [inline data_model/src/block.rs:193]: ok, let's leave this one

File: data_model/src/block.rs
Comment (mversic) [inline data_model/src/block.rs:178]: do you think we should expose that blocks have headers? or should we do the same as for the `payload`?

Comment (Stukalov-A-M) [reply data_model/src/block.rs:178]: There are 18 places where the headers are used. If we hide the method, we should define a versioned SignedBlock on each of these place. And my thoughts hint me that it would influence on the performance and the convenience of the code maintenance.

Comment (mversic) [reply data_model/src/block.rs:178]: what I meant was this: instead of having `block.header().height` we can have `block.height()`. Likewise for all other header fields

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:376]:
```suggestion
    #[cfg(feature = "std")]
    #[cfg(feature = "transparent_api")]
    pub fn hash_of_payload(&self) -> iroha_crypto::HashOf<TransactionPayload> {
```

Comment (mversic) [reply data_model/src/transaction.rs:376]: I suspect this method can even be removed. It's probably unused

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:310]:
```suggestion
    #[cfg(feature = "std")]
    #[cfg(feature = "transparent_api")]
    pub fn hash_of_payload(&self) -> iroha_crypto::HashOf<TransactionPayload> {
```

File: data_model/src/block.rs
Comment (mversic) [inline data_model/src/block.rs:215]:
```suggestion
    #[cfg(feature = "std")]
    #[cfg(feature = "transparent_api")]
    pub fn hash_of_payload(&self) -> iroha_crypto::HashOf<BlockPayload> {
```

File: data_model/src/block.rs
Comment (Stukalov-A-M) [reply data_model/src/block.rs:178]: It's uses in few places, so I decided to keep it.
But the getters of else BlockHeader fields I've implemented in the SignedBlock

File: data_model/src/transaction.rs
Comment (Stukalov-A-M) [reply data_model/src/transaction.rs:376]: No, it's used

Comment (Stukalov-A-M) [reply data_model/src/transaction.rs:376]: It's used in the Client crate 
image

Comment (mversic) [reply data_model/src/transaction.rs:376]: that's not the one defined on `TransactionValue`

File: data_model/src/block.rs
Comment (mversic) [reply data_model/src/block.rs:178]: >  But the getters of else BlockHeader fields I've implemented in the SignedBlock

please revert that

File: data_model/src/transaction.rs
Comment (mversic) [reply data_model/src/transaction.rs:376]: you can remove this one

Comment (Stukalov-A-M) [reply data_model/src/transaction.rs:376]: Ok. This one is defined in the SignedTransaction and is used:
image

And this one is defined in the TransactionValue and is used as well:
image

All of these method outputs were tightly connected to a payload() method that was removed. So i can not remove any of newly created method without a deep refactoring all stable logic.

Comment (Stukalov-A-M) [reply data_model/src/transaction.rs]: Added a transparent API macro

File: data_model/src/block.rs
Comment (Stukalov-A-M) [reply data_model/src/block.rs:180]: Reverted all getters back. Kept headers() only

**Review by 6r1d — APPROVED** [review]
LGTM

File: data_model/src/transaction.rs
Comment (Arjentix) [inline data_model/src/transaction.rs:104]: Isn't it fixed?

File: data_model/src/transaction.rs
Comment (Arjentix) [inline data_model/src/transaction.rs:121]: Same

**Review by Arjentix — COMMENTED** [review]
LGTM, but I'd wait for Marin's approval

File: data_model/src/transaction.rs
Comment (mversic) [reply data_model/src/transaction.rs:104]: no

File: core/src/block.rs
Comment (mversic) [reply core/src/block.rs:249]: > Hm, we could implement Deref<Target=SignedBlock> to not to copy-paste these methods... Or make a getter for .0

`Deref` is for smart pointers. We can implement `AsRef`

File: data_model/src/block.rs
Comment (mversic) [inline data_model/src/block.rs:148]:
```suggestion
```

this seems useful to the end user

File: data_model/src/block.rs
Comment (mversic) [inline data_model/src/block.rs:188]:
```suggestion
```

also

File: data_model/src/query/mod.rs
Comment (mversic) [inline data_model/src/query/mod.rs:264]:
```suggestion
```

IMO, I would just remove these

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:394]: we should consider removing all methods defined on `TransactionValue`. I'm not sure what desired api would be here

Comment (mversic) [reply data_model/src/transaction.rs:394]: we can remove all methods on `TransactionValue` and leave only `fn error(&self)` and implement `AsRef<SignedTransaction>`

File: core/src/block.rs
Comment (mversic) [inline core/src/block.rs:448]:
```suggestion
            if !self.header().is_genesis() {
```

File: core/src/block.rs
Comment (mversic) [inline core/src/block.rs:503]:
```suggestion
            if !self.header().is_genesis()
```

File: core/src/block.rs
Comment (mversic) [inline core/src/block.rs:513]:
```suggestion
            if self.header().is_genesis() {
```

File: core/src/block.rs
Comment (mversic) [inline core/src/block.rs:564]: this is often repeated. Make it into a function. It's ok to have `payload` function in this case

File: core/src/block.rs
Comment (mversic) [reply core/src/block.rs:670]: yes, much better

File: core/src/block.rs
Comment (mversic) [inline core/src/block.rs:672]: remove all methods and implement `AsRef`

File: core/src/block.rs
Comment (Stukalov-A-M) [reply core/src/block.rs:564]: Disagree. It repeats only in tests and only 4 times. I don't think it's a good idea to make a function only for testing

Comment (mversic) [reply core/src/block.rs:564]: @Erigara can you share your opinion here?

File: data_model/src/transaction.rs
Comment (Stukalov-A-M) [reply data_model/src/transaction.rs:394]: What do you mean by that `fn error(&self)` ?

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:352]: how will a user get `TransactionValue::error`? you have to make a getter function for that field

File: data_model/src/query/mod.rs
Comment (mversic) [inline data_model/src/query/mod.rs:270]: I cannot comment in the right place, but please change struct definition to:
```rs
    /// Output of [`FindAllTransactions`] query
    #[derive(
        Debug, Clone, PartialEq, Eq, Getters, Decode, Encode, Deserialize, Serialize, IntoSchema,
    )]
    #[getset(get = "pub")]
    #[ffi_type]
    pub struct TransactionQueryOutput {
        /// The hash of the block to which `tx` belongs to
        pub block_hash: HashOf<SignedBlock>,
        /// Transaction
        #[getset(skip)]
        pub transaction: Box<TransactionValue>,
    }
```

File: data_model/src/transaction.rs
Comment (Stukalov-A-M) [reply data_model/src/transaction.rs:352]: Added `[getset]` macro

Comment (Stukalov-A-M) [reply data_model/src/transaction.rs:354]: Added `[getset]` macro

File: core/src/block.rs
Comment (mversic) [reply core/src/block.rs:530]: > Disagree. It repeats only in tests and only 4 times. I don't think it's a good idea to make a function only for testing

I see your point now. How about you define a freestanding function in the `test` module instead of a method

**Review by mversic — APPROVED** [review]

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:369]: sorry, I just noticed this is inconsistent. Remove `PartialOrd` and `Ord` manual impls and derive them

File: data_model/src/query/mod.rs
Comment (mversic) [inline data_model/src/query/mod.rs:261]: sorry, I just noticed this is inconsistent. Remove `PartialOrd` and `Ord` manual impls and derive them

**Review by mversic — APPROVED** [review]

**Review by 6r1d — APPROVED** [review]

**Pull request merged**
